pkg/target/http.go

```diff
@@ -362,25 +392,78 @@ func (ht *HTTPTarget) Write(messages []*models.Message) (*models.TargetWriteResu
 				continue
 			}
 			defer resp.Body.Close()
-			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+			responseBody, err := io.ReadAll(resp.Body)
+
+			if err != nil {
+				failed = append(failed, goodMsgs...)
+				errResult = multierror.Append(errResult, errors.New("Error reading response body: "+err.Error()))
+				continue
+			}
+
+			response := response{Body: string(responseBody), Status: resp.StatusCode}
+
+			if response.Status >= 200 && response.Status < 300 {
 				for _, msg := range goodMsgs {
 					if msg.AckFunc != nil { // Ack successful messages
 						msg.AckFunc()
 					}
 					sent = append(sent, msg)
 				}
-			} else {
-				errResult = multierror.Append(errResult, errors.New("Got response status: "+resp.Status))
-				failed = append(failed, goodMsgs...)
 				continue
 			}
+
+			if findMatchingRule(response, ht.responseRules.Invalid) != nil {
+				invalid = append(invalid, goodMsgs...)
+				continue
+			}
+
+			if findMatchingRule(response, ht.responseRules.SetupError) != nil {
+				hitSetupError = true
+			}
+
+			errResult = multierror.Append(errResult, fmt.Errorf("Got response status: '%s' with body '%s'", resp.Status, response.Body))
```

Comment:
We must be careful to either sanitise, omit the body, or never log this - respose bodies can contain values.

Also note - when you hit some http endpoints you get a whole html page response, which is inconvenient too

Reply:
Yes, that's true, it's risky to use it. But also body with some error code/message would be also the best place to find info what's wrong.

There is this option now to define string which should be present in the response body to make a invalid/setup rule match...maybe we could use that instead of the whole body?

Reply:
Yup, agreed - I don't like that currently we only get the code.

> There is this option now to define string which should be present in the response body to make a invalid/setup rule match...maybe we could use that instead of the whole body?

That's likely better for certain, but not perfect. Is there a scenario where we won't have a match but would like to log the error? If so what would we do then?

Reply:
> s there a scenario where we won't have a match but would like to log the error? If so what would we do then?

In this case..I think we should still rely on http status only, so what we have now. In the future maybe we can add some safe/sanitized response provider 🤔 So how about for now:

* for invalid match - use full response body, which will be a part of bad row
* for setup match - use configured part of response body, we know it's safe (this part of a log and later part of an alert)
* for no match, so all transient - only http status, same we have now

WDYT?

Reply:
Yeah that'll work